File: back-end/po-service/src/metrics-domain.js
// PO service domain metrics.
//   po_state_transitions_total{from,to,outcome}  — counter
//   po_approval_duration_seconds                 — histogram (for SLO)
//   po_cross_service_calls_total{target,status}  — counter

const { metrics } = require('@opentelemetry/api');

const meter = metrics.getMeter('gep-po-domain');

const stateTransitionsTotal = meter.createCounter('po_state_transitions_total', {
  description: 'Purchase-order state transitions. Labels from, to, outcome.',
});

const approvalDuration = meter.createHistogram('po_approval_duration_seconds', {
  description: 'Time between PO creation and final approval, in seconds.',
  unit: 's',
});

const crossServiceCalls = meter.createCounter('po_cross_service_calls_total', {
  description: 'Outbound calls to IAM / Supplier. Labels target, status.',
});

module.exports = {
  recordTransition: (from, to, outcome = 'ok') =>
    stateTransitionsTotal.add(1, { from, to, outcome }),
  recordApprovalDuration: (seconds) => approvalDuration.record(seconds),
  recordCrossServiceCall: (target, status) =>
    crossServiceCalls.add(1, { target, status: String(status) }),
};

File: back-end/supplier-service/app/metrics_domain.py
"""Supplier service domain metrics.

  supplier_blacklist_hits_total          - counter
  supplier_crud_total{op,status}         - counter
  supplier_state_transitions_total{from,to,outcome}  - counter
"""
from __future__ import annotations

from opentelemetry import metrics as otel_metrics

_meter = otel_metrics.get_meter("gep-supplier-domain")

blacklist_hits_total = _meter.create_counter(
    name="supplier_blacklist_hits_total",
    description="Number of blacklist check hits (supplier matched blacklist).",
)

crud_total = _meter.create_counter(
    name="supplier_crud_total",
    description="CRUD operation count. Labels op, status.",
)

state_transitions_total = _meter.create_counter(
    name="supplier_state_transitions_total",
    description="Supplier lifecycle state transitions. Labels from, to, outcome.",
)


def record_blacklist_hit(supplier_id: str | None = None) -> None:
    blacklist_hits_total.add(1, {"supplier_id": supplier_id or "unknown"})


def record_crud(op: str, status: str) -> None:
    crud_total.add(1, {"op": op, "status": status})


def record_transition(frm: str, to: str, outcome: str = "ok") -> None:
    state_transitions_total.add(1, {"from": frm, "to": to, "outcome": outcome})
